### PR TITLE
GH-35675 [C++] Don't copy the ArraySpan into the REE ArraySpan

### LINKED ISSUE #35675: [C++] Copying a REE ArraySpan that has buffers using scratch buffers could lead to memory errors
### Describe the bug, including details regarding any error messages, version, and platform.

More details in #35581. To avoid this risk, I want to also avoid copying `ArraySpan` into `RunEndEncodedArraySpan` when it's instantiated.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by felipecrv**
### Rationale for this change

Copying ArraySpan is not without issues (see #35581), and this change seems good by itself as it makes it easier for the compiler to SROA [1] the whole `RunEndEncodedArraySpan` class.

[1] https://www.llvm.org/docs/Passes.html#sroa-scalar-replacement-of-aggregates

### What changes are included in this PR?
 
 - Make `array_span` private in `ree_util::RunEndEncodedArraySpan`
 - Allow construction of  `ree_util::RunEndEncodedArraySpan` with separate `offset` and `length`
 - Change `RunEndEncodedBuilder` to avoid a copy of the `ArraySpan` being iterated over
 - Prevent instantiation based on implicit conversion from `ArrayData` to `ArraySpan`
 
### Are these changes tested?

Yes, by the existing unit tests of the various uses of `RunEndEncodedArraySpan`.

### Are there any user-facing changes?

No meaningful changes other than some small tweaks in the interface of a recently added class.

* Closes: #35675

### CODE REVIEW

File: cpp/src/arrow/util/ree_util.h [hunk 1]
```diff
@@ -227,20 +227,39 @@ class RunEndEncodedArraySpan {
...
-    assert(array_span.type->id() == Type::RUN_END_ENCODED);
+  /// \brief Construct a RunEndEncodedArraySpan from an ArraySpan and new
+  /// absolute offset and length.
+  ///
+  /// RunEndEncodedArraySpan{span, off, len} is equivalent to:
+  ///
+  ///   span.SetSlice(off, len);
+  ///   RunEndEncodedArraySpan{span}
+  ///
+  /// ArraySpan::SetSlice() updates the null_count to kUnknownNullCount, but
```
Comment (bkietz) [inline cpp/src/arrow/util/ree_util.h:242]: Could this be fixed by making SetSlice() update null_count differently? If the original span has zero nulls, then shouldn't the slice also have zero nulls? See for example https://github.com/apache/arrow/pull/35628/files#diff-c5d8f67626f30c66a79d1f677ee81264a9b044280e942cfa846e07306b25365bR478

Comment (felipecrv) [reply cpp/src/arrow/util/ree_util.h:242]: That improvement seems unrelated to this change and affects code that is much more widely used. Could you create an issue for that? My comment here is just an informal proof that the equivalence here holds even though I don't perform all the operations that `SetSlice` does based on my class only taking ArraySpans of REEs.

**Review by bkietz — APPROVED** [review]
LGTM

**Comment by felipecrv** [comment]
@bkietz I rebased and force-pushed to see if CI passes now.

**Comment by felipecrv** [comment]
One test is timing out. Can someone re-trigger it instead of me pushing again?  cc @bkietz @pitrou

**Comment by felipecrv** [comment]
@pitrou this is ready for merging.

**Review by pitrou — APPROVED** [review]

**Pull request merged**